lib/puppet/application/lookup.rb

```diff
@@ -49,8 +60,49 @@ class Puppet::Application::Lookup < Puppet::Application
   def run_command
     options[:keys] = command_line.args
 
-    puts "Hello World!"
-    puts options
+    if options[:keys].empty?
+     raise "No keys were given to lookup."
+    end
+
+    if !options[:node]
+      raise "No node was given via the '--node' flag for the scope of the lookup."
+    end
+
+    if (options[:sort_merge_arrays] || options[:merge_hash_arrays] || options[:prefix] || options[:unpack_arrays]) && options[:merge] != 'deep'
+      raise "The options #{DEEP_MERGE_OPTIONS} are only available with '--merge deep'\n#{RUNHELP}"
+    end
+
+    scope = generate_scope
+
+    use_default_value = !options[:default_value].nil?
+    merge_options = nil
+
+    if options[:merge]
+      if options[:merge] == 'deep'
+        merge_options = {'strategy' => 'deep',
+          'sort_merge_arrays' => !options[:sort_merge_arrays].nil?,
+          'merge_hash_arrays' => !options[:merge_hash_arrays].nil?}
+
+        if options[:prefix]
+          merge_options.merge({'prefix' => options[:prefix]})
+        end
+
+        if options[:unpack_arrays]
+          merge_options.merge({'unpack_arrays' => options[:unpack_arrays]})
+        end
+
+      else
+        merge_options = {'strategy' => options[:merge]}
+      end
+    end
+
+    puts Puppet::Pops::Lookup.lookup(scope, options[:keys], options[:type], options[:default_value], use_default_value, {}, {}, merge_options)
   end
 
+  def generate_scope
+    node = Puppet::Node.indirection.find("#{options[:node]}")
+    compiler = Puppet::Parser::Compiler.new(node)
+    compiler.compile
```

Comment:
There are other problems with the approach of compiling the catalog to lookup. Say your compilation fails because you have some problems with your data configuration and you want to use lookup with -explain to understand what is going on. If a compile is required it will fail, and you will not be able to lookup. While being useful to do the compile it is also somewhat different from how it will actually work at runtime since it will first compile everything, and then perform the lookup. In practice the lookups while take place in the middle of the evaluation somewhere (where some of the global variables may not yet have been evaluated). Also, when using -explain, if compiling the entire catalog, there will be many many calls to lookup things - should all of them be explained? 

Also, regarding speed, it is not reasonable to perform a lengthy compilation (a real case could take many minutes) to just lookup a value. 

For the default case we do need the compiler, and that the infrastructure is initialized (the environment instance, injectors etc. and that calls have been made to set_node_parameters, and create_settings_scope. For other variables than facts and settings it would be better if the user had to provide them directly (i.e as overrides) on the command line. This way you can test "what will be looked up if these variables are set this way....", then you know your data configuration is correct.